instrumentation/grpc/src/main/java/brave/grpc/TracingClientInterceptor.java

```diff
@@ -30,95 +32,158 @@
-
-            // See RATIONALE.md which notes response callbacks are in the invocation context
-            responseListener = new TraceContextCallListener<>(
-              responseListener,
-              currentTraceContext,
-              invocationContext
-            );
-
-            super.start(new TracingClientCallListener<>(responseListener, span), headers);
-          }
-        }
-
-        @Override public void sendMessage(ReqT message) {
-          try (Scope scope = currentTraceContext.maybeScope(span.context())) {
-            super.sendMessage(message);
-            parser.onMessageSent(message, span.customizer());
-          }
-        }
-      };
-    } catch (Throwable e) {
-      error = e;
-      throw e;
-    } finally {
-      if (error != null) span.error(error).finish();
+    return new TracingClientCall<>(
+      method, callOptions, currentTraceContext.get(), next.newCall(method, callOptions));
+  }
+
+  void finish(GrpcClientResponse response, @Nullable Span span) {
+    if (span == null || span.isNoop()) return;
+    Throwable error = response.error();
+    if (error != null) span.error(error);
+    parser.onClose(response.status, response.trailers, span.customizer());
+    span.finish();
+  }
+
+  void finishWithError(@Nullable Span span, Throwable error) {
+    if (span == null || span.isNoop()) return;
+    if (error != null) span.error(error);
+    span.finish();
+  }
+
+  final class TracingClientCall<ReqT, RespT> extends SimpleForwardingClientCall<ReqT, RespT> {
+    final MethodDescriptor<ReqT, RespT> method;
+    final CallOptions callOptions;
+    final TraceContext invocationContext;
+    final AtomicReference<Span> spanRef = new AtomicReference<>();
+
+    TracingClientCall(MethodDescriptor<ReqT, RespT> method, CallOptions callOptions,
+      TraceContext invocationContext, ClientCall<ReqT, RespT> call) {
+      super(call);
+      this.method = method;
+      this.callOptions = callOptions;
+      this.invocationContext = invocationContext;
+    }
+
+    @Override public void start(Listener<RespT> responseListener, Metadata headers) {
+      GrpcClientRequest request =
+        new GrpcClientRequest(nameToKey, method, callOptions, delegate(), headers);
+
+      Span span = tracer.nextSpanWithParent(sampler, request, invocationContext);
+      injector.inject(span.context(), request);
+      if (!span.isNoop()) {
+        span.kind(Span.Kind.CLIENT).start();
+        parser.onStart(method, callOptions, headers, span.customizer());
+      }
+      spanRef.set(span);
+
+      responseListener = new TracingClientCallListener<>(
+        responseListener,
+        invocationContext,
+        spanRef,
+        request
+      );
+
+      try (Scope scope = currentTraceContext.maybeScope(span.context())) {
+        super.start(responseListener, headers);
+      } catch (Throwable e) {
+        finishWithError(spanRef.getAndSet(null), e);
```

Comment:
I think close will be called with this error packed into the `Status` and we probably don't need to specially handle it.

Reply:
userInterceptor_throwsOnStart now exists to explain this.